### PR TITLE
Add support for global flags

### PR DESCRIPTION
**Pull request by lcmen**
Hi @mrrooijen ,
Thank you for `commander`. It really helps with building nice interfaces for CLI applications.

I've just started learning Crystal and I would like to build small CLI app. I wanted to use `commander` for defining interface of my program but I've noticed there is no support for global flags. It makes sub-command execution a little bit strange: `s3 push -config ~/.config file.png` vs `s3 -config ~/.config push file.png`.

I've made PoC for adding global flags which are can be consumed by sub-commands. However there is one case which I don't know how to solve - raising exception if flag is not defined. It's caused by the fact that we don't know whether the flag is defined in sub-commands when we parse parent command's flags.

Since I'm new to Crystal I'm pushing my WIP for initial review and I'm looking forward to your feedback!

### CODE REVIEW

File: src/commander/option.cr [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  def key
+    flag.name
+  end
+
+  def global?
+    !!flag.global
+  end
+
+  def self.build(value, flag)
```
Comment (lcmen) [inline src/commander/option.cr:14]: I've added this helper method as using `Option.new` didn't correctly recognize the type. I was getting a following error:

```
Error: no overload matches 'Commander::Options#set' with types String, Commander::Option(Bool | Float32 | Float64 | Int32 | Int64 | String | Nil)
```

File: src/commander/parser/short_flag_format.cr [hunk 2]
```diff
@@ -36,7 +38,7 @@ class Commander::Parser::ShortFlagFormat < Commander::Parser::Base
         end
       end
 
-      no_such_flag!(flag_char)
```
Comment (lcmen) [inline src/commander/parser/short_flag_format.cr:39]: This code affects the issue I mentioned in the PR.

**Comment by mrrooijen** [comment]
Hey,

Thanks for the kind words and effort.

I deliberately chose to have a fixed order in which to execute commands:

```
program [(sub)command] [flags] [args]
```

This was mainly to enforce consistency and maintain simplicity. That being said, I've wanted to implement persistent flags for a while using simple recursion.

See: https://github.com/mrrooijen/commander/commit/50236aa610db9e45a6d5575bee80f52b8d6d0346

This doesn't change the positioning (so it doesn't accomplish what you were attempting) but it allows flags that you want available on each child command to become available automatically without having to redefine or manually share the flag definition. For global flags you'd just define a persistent flag at the root, and all commands will inherit it.

I don't think that I want to add a special global variable just for the purpose of being able to position flags before any (sub)commands, though.

**Comment by lcmen** [comment]
Thanks @mrrooijen for your feedback. I understand your motives.

Maybe I'm biased because I'm kind of a terminal guy and I favour global flag before the sub-commands. In this case I will look for some alternative library.

Anyway it was nice exercise to improve my Crystal's skills :).

**Comment by mrrooijen** [comment]
Sure thing. Note that my motive might not conform to the norm. I'm actually not hardcore when it comes to the terminal, and I definitely appreciate your input. I used [Cobra](https://github.com/spf13/cobra) in the past and wanted something similar to use in Crystal, hence Commander. Perhaps I'll eventually end up adding global flag support, but not sure at this time.

**Pull request closed without merging**